Sources/LookupParser.swift

```diff
@@ -0,0 +1,104 @@
+import Foundation
+
+/// A structure used to represent a template variable, and to resolve it in a given context.
+struct LookupParser {
+  private var bits = [String]()
+  private var current = ""
+  private var partialBits = [String]()
+  private var referenceLevel = 0
+
+  let variable: String
+  let context: Context
+
+  // Split the lookup string and resolve references if possible
+  init(_ variable: String, in context: Context) {
+    self.variable = variable
+    self.context = context
+  }
+
+  mutating func parse() throws -> [String] {
+    defer {
+      bits = []
+      current = ""
+      partialBits = []
+      referenceLevel = 0
+    }
+
+    for c in variable.characters {
+      switch c {
+      case "." where referenceLevel == 0:
+        try foundSeparator()
+      case "[":
+        try openBracket()
+      case "]":
+        try closeBracket()
+      default:
+        current.append(c)
+      }
+    }
+    try finish()
+
+    return bits
+  }
+
+  private mutating func foundSeparator() throws {
+    if !current.isEmpty {
+      partialBits.append(current)
+    }
+
+    guard !partialBits.isEmpty else {
+      throw TemplateSyntaxError("Attempting to dereference empty object in variable '\(variable)'")
```

Comment:
It's more than just `..`, it could be `stuff..`, or `stuff..other.stuff`, or other cases. It's already a `TemplateSyntaxError`, this just specifies what's wrong.

Reply:
If I understand code flow correctly here errors will all fall into `unexpected '.' in \(variable)` kind of description.